static/src/javascripts/projects/commercial/modules/messenger/scroll.js

```diff
@@ -0,0 +1,154 @@
+// @flow
+import { addEventListener } from 'lib/events';
+import detect from 'lib/detect';
+import fastdom from 'lib/fastdom-promise';
+import { register } from 'commercial/modules/messenger';
+
+// An intersection observer will allow us to efficiently send slot
+// coordinates for only those that are in the viewport.
+const w = window;
+let useIO = 'IntersectionObserver' in w;
+let taskQueued = false;
+let iframes: {
+    [iframeId: string]: any,
+} = {};
+let iframeCounter: number = 0;
+let observer;
+let visibleIframeIds;
+
+const reset = (useIO_: boolean): void => {
+    useIO = useIO_;
+    taskQueued = false;
+    iframes = {};
+    iframeCounter = 0;
+};
+
+type DOMRect = window.DOMRect;
+type IframeTuple = [string, DOMRect];
+type FrameCoords = {
+    width: number,
+    height: number,
+    top: number,
+    bottom: number,
+    left: number,
+    right: number,
+};
+
+// Instances of classes bound to the current view are not serialised correctly
+// by JSON.stringify. That's ok, we don't care if it's a DOMRect or some other
+// object, as long as the calling view receives the frame coordinates.
+const domRectToRect = (rect: DOMRect): FrameCoords => ({
+    width: rect.width,
+    height: rect.height,
+    top: rect.top,
+    bottom: rect.bottom,
+    left: rect.left,
+    right: rect.right,
+});
+
+const sendCoordinates = (iframeId: string, domRect: DOMRect): void => {
+    iframes[iframeId].respond(null, domRectToRect(domRect));
+};
+
+const getDimensions = (id: string): IframeTuple => [
+    id,
+    iframes[id].node.getBoundingClientRect(),
+];
+
+const isIframeInViewport = function(item: IframeTuple): boolean {
+    return item[1].bottom > 0 && item[1].top < this.height;
```

Comment:
it returns `true` as soon as either the bottom part enters the top of the viewport and until the top part leaves the bottom of the viewport. it works because `bottom < top` is a tautology.